### PR TITLE
Adding rust tests for match boundaries

### LINKED ISSUE #61894: Port rust regex tests
Rust has lots of regex functional tests we should port over to improve our coverage:
https://github.com/rust-lang/regex/tree/5197f21287344d2994f9cf06758a3ea30f5a26c3/tests

### PR DESCRIPTION
**Pull request by joperezr**
Fixes https://github.com/dotnet/runtime/issues/61894

This is a port of the tests at https://github.com/rust-lang/regex/tree/6ff285e37555d4adc52f24e97318681f8a5ecd48/tests. Particularly I'm porting all of the patterns and inputs that are using the `matitre` and `mat` macros which are checking for match boundaries for given inputs. There were some tests that couldn't be ported since they relied on syntax that wasn't supported by .NET.

cc: @stephentoub

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(MatchStartAndEndPositionsTestData))]
+        public async Task MatchStartAndEndPositions(RegexEngine engine, string pattern, string input, IEnumerable<(int, int)>? matchBoundaries)
+        {
+            Regex regex = await RegexHelpers.GetRegexAsync(engine, pattern);
+            Match match = regex.Match(input);
+            if (matchBoundaries == null)
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs:426]: Where is it null?

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+                yield return new object[] { engine, @"(?m)^$", "abc\ndef\nxyz", new ValueTuple<int, int>[] { } };
+                yield return new object[] { engine, @"(?m)^", "abc\ndef\nxyz", new[] { (0, 0), (4, 4), (8, 8) } };
+                yield return new object[] { engine, @"(?m)$", "abc\ndef\nxyz", new[] { (3, 3), (7, 7), (11, 11) } };
+                yield return new object[] { engine, @"(?m)^[a-z]", "abc\ndef\nxyz", new[] { (0, 1), (4, 5), (8, 9) } };
+                yield return new object[] { engine, @"(?m)[a-z]^", "abc\ndef\nxyz", new ValueTuple<int, int>[] { } };
+                yield return new object[] { engine, @"(?m)[a-z]$", "abc\ndef\nxyz", new[] { (2, 3), (6, 7), (10, 11) } };
+                yield return new object[] { engine, @"(?m)$[a-z]", "abc\ndef\nxyz", new ValueTuple<int, int>[] { } };
+                yield return new object[] { engine, @"(?m)^$", "", new[] { (0, 0) } };
+                yield return new object[] { engine, @"(?m)(?:^$)*", "a\nb\nc", new[] { (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5) } };
+                if (engine != RegexEngine.NonBacktracking) // https://github.com/dotnet/runtime/issues/72470
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs:64]:
```suggestion
                if (!RegexHelpers.IsNonBacktracking(engine)) // https://github.com/dotnet/runtime/issues/72470
```

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+                yield return new object[] { engine, @"(a|b)c*d", "abcd", new[] { (1, 4) } };
+                yield return new object[] { engine, @"(ab|ab*)bc", "abc", new[] { (0, 3) } };
+                yield return new object[] { engine, @"a([bc]*)c*", "abc", new[] { (0, 3) } };
+                yield return new object[] { engine, @"a[bcd]*dcdcde", "adcdcde", new[] { (0, 7) } };
+                yield return new object[] { engine, @"(ab|a)b*c", "abc", new[] { (0, 3) } };
+                yield return new object[] { engine, @"[A-Za-z_][A-Za-z0-9_]*", "alpha", new[] { (0, 5) } };
+                yield return new object[] { engine, @"^a(bc+|b[eh])g|.h$", "abh", new[] { (1, 3) } };
+                yield return new object[] { engine, @"abcd", "abcd", new[] { (0, 4) } };
+                yield return new object[] { engine, @"a(bc)d", "abcd", new[] { (0, 4) } };
+                yield return new object[] { engine, @"a[☺-♥]?c", "a☻c", new[] { (0, 3) } };
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs:308]: Can we use \uxxxx instead of non-ASCII characters?

**Review by stephentoub — APPROVED** [review]
Nice. Thanks.

**Comment by stephentoub** [comment]
> Particularly I'm porting all of the patterns and inputs that are using the matitre and mat macros which are checking for match boundaries for given inputs

This PR states that it fixes the rust test porting.  Are there no tests beyond these?

**Comment by joperezr** [comment]
> Are there no tests beyond these worth looking into?

There are some isMatch tests but I believe that for all of those they had a corresponding match boundaries test that I did port. They also have some Replace tests which I didn't port since they were using different syntax for the replacement's catpuring groups backreferences so it would be hard to point each test manually. They also have some unit tests around their DFA engine which are not really applicable, since they are not full scenario tests and just inspect internal state. Finally they have few split tests as well which I didn't port mostly because I believe we already had coverage for those.

So, to answer your question, I think we can probably find a few more interesting tests from them to port, but I expect that the ones in this PR are at least most of the ones we care about.

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs (hunk 3 above)
Comment (joperezr) [reply src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs:308]: Yup, I thought I had built my reader to automatically convert those, not sure why they weren't converted here.

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs (hunk 1 above)
Comment (joperezr) [reply src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexRustTests.cs:426]: It's not, good point. I had initially made this parameter nullable and would pass null for no matches, but I switched that so I'll remove this if condition.

**Comment by joperezr** [comment]
I forgot to add stats, but this is roughly adding 1600 new tests, which locally take ~20 seconds to run, so this brings the total test execution time for regex tests from ~40 seconds to ~1 minute in my machine.

**Comment by stephentoub** [comment]
> I expect that the ones in this PR are at least most of the ones we care about.

Ok, great, thanks

**Comment by danmoseley** [comment]
Do we need any addition to third party notices file?

**Comment by joperezr** [comment]
Good question, I suppose the answer is yes, if so I can send a follow up PR. @richlander is that the case?

**Comment by richlander** [comment]
We have a simple pattern:

- Affects product -> add to root TPN.
- Affects tests -> add to localized TPN, like we've done for perf tests -- https://github.com/dotnet/runtime/blob/main/src/tests/JIT/Performance/CodeQuality/Burgers/THIRD-PARTY-NOTICES

**Pull request merged**